### PR TITLE
`DateTimePicker`: fix date format when switching to 12-hour time format

### LINKED ISSUE #17970: TimePicker is12hour format is reversed
**Describe the bug**
The order of inputs for a Timepicker is reversed. 
In the documentation is described that the dateformat is assumed to be mm-dd-yyyy if is12Hour is true, currently the order is switched on `is12Hour`.

**To reproduce**
 1. Insert Timepicker component in Block
 2. set is12Hour to true

or
here is the responsible function https://github.com/WordPress/gutenberg/blob/master/packages/components/src/date-time/time.js#L243-L247

**Expected behavior**
```
is12Hour == true => MM-DD-YYYY
is12Hour == false => DD-MM-YYYY
```

**Screenshots**
block editor
code snippet

**Desktop (please complete the following information):**
 - OS: all
 - Browser all
 - Version all

**Smartphone (please complete the following information):**
 - Device: all
 - OS: all
 - Browser: all
 - Version: all

**Additional context**
- Tested in: WordPress 5.2.4

### PR DESCRIPTION
**Pull request by amustaque97**
Fixes: #17970

## Description

Updated the `TimePicker` component to show time in `MM-DD-YYYY` format as mentioned in the [docs](https://developer.wordpress.org/block-editor/reference-guides/components/date-time/#is12hour) 
## How has this been tested?

- Run storybook
- Navigate to `TimePicker` component
- Under the knob section select 12-hours format
- Verify the format
## Screenshots 
image

## Types of changes

Bugfix
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by paaljoachim** [comment]
Thank you very much for working on this feature, @amustaque97 
Let's see I believe @adamsilverstein and @retrofox both have worked on the date picker. 
Perhaps Adam og Damian could take a closer look at this PR.

**Comment by ciampo** [comment]
Hey @amustaque97 , thank you for looking into this!

You definitely spotted an inconsistency between what is documented and how the component behaves. I'm not sure, though, that the fix proposed here is right one.

The docs for the `is12Hour` prop mention:

> Whether we use a 12-hour clock. With a 12-hour clock, an AM/PM widget is displayed and the time format is assumed to be `MM-DD-YYYY`.

Judging by how the code was written (including tests written specifically for that condition) and by this sentence, I'm inclined to think that the desired behavior here would be to:

- by default, when `is12Hour` is `false` (i.e a 24h format is being used), the date format is `DD-MM-YYYY`
- when `is12Hour` is set to `true`, the date format changes to `MM-DD-YYYY` 

In this perspective, I would say the the behavior of the component is currently the opposite of what we'd like it to be.

My inclination here is to:
- implement the `MM-DD-YYYY` format when `is12Hour` is `true`, and otherwise use the `DD-MM-YYYY` format when `is12Hour` is `false` (it looks like this could be done just by flipping the ternary logic used to compute `dayMonthFormat`)
- improve the docs for the `is12Hour` prop to be more explicit, and therefore cause less confusion

**Comment by amustaque97** [comment]
@ciampo thank you so much for taking time to review my PR 😄 .
How should we go about it now? I once again went through the [slack discussion](https://wordpress.slack.com/archives/C02QB2JS7/p1625548800325600) and [docs](https://developer.wordpress.org/block-editor/reference-guides/components/date-time/#is12hour) but not able to finalise things.

**Comment by ciampo** [comment]
> @ciampo thank you so much for taking time to review my PR 😄 . How should we go about it now? I once again went through the [slack discussion](https://wordpress.slack.com/archives/C02QB2JS7/p1625548800325600) and [docs](https://developer.wordpress.org/block-editor/reference-guides/components/date-time/#is12hour) but not able to finalise things.

I was not aware of the slack conversation, where [it was also assumed that currently, the bug is caused by the logic being the other way around of what it should be](https://wordpress.slack.com/archives/C02QB2JS7/p1625549504332300).

It was also discussed that tying the date format (`DD-MM-YYYY` vs `MM-DD-YYYY`) to the time format (`12` vs `24` hours) is not ideal — and I agree! I personally think that the two formats should be chosen and matched independently from each other.

So, what next?

We could remove the `is12Hour` check (similarly to what has been done so far in this PR) and introduce a new prop to allow consumers of the component to change the date format — although we'd first need to define a couple of things:

- _what would be the default date format?_  To this question, I'd argue that the default date format should be the same that is intended for the component to have by default today — which is `DD-MM-YYYY`
- _what values should the new prop assume?_ This is the hard part. I believe that it would be wrong to assume that the only allowed values should be `DD-MM-YYYY` or `MM-DD-YYYY`(for example, what about `YYYY-MM-DD`). We should, therefore, probably have allow a set of possible values to choose from.

A simpler approach would be to merge a quick fix in this PR (switch the logic of the `is12Hour` check), and tackle the additional of a new prop to format the date in a separate PR.

What do folks think?

**Comment by amustaque97** [comment]
Thanks for the input 😄 . 

> what would be the default date format?
`DD-MM-YYYY` 👍🏻 

> what values should the new prop assume?
I would personally recommend to allow these types are currently being used in WordPress
Screenshot 2022-01-25 at 8 28 48 PM


To get other people input I can raise this concern in our Wednesday meeting!

**Comment by ciampo** [comment]
> I would personally recommend to allow these types are currently being used in WordPress

What you're referring to are [these date formatting settings / parameters](https://wordpress.org/support/article/formatting-date-and-time/). Here's a few personal considerations:

- while it's great to have consistency in our date formats, the format that you suggested is globally used in the PHP language, and not as much in JS.
- we would also need to understand how much work it would require to implement such date formatting logic (also considering that today the component uses `moment.js`, but this may change in an effort to shave some bytes off the overall library size)

In the light of the above, I would recommend to split the work in 2 parts:

- this PR should focus on fixing the bug (i.e. inverting the `is12Hour` check). This ensures that the fix is released as soon as possible, and is not delayed by any further discussion around the date format.
- we should open a separate issue to discuss the addition of a "date format" prop. If you'd like to bring the topic up during the next meeting, we'll be able to link to that issue for folks to chime in and discuss

**Comment by amustaque97** [comment]
> I would recommend to split the work in 2 parts

Sure, make sense.

> this PR should focus on fixing the bug (i.e. inverting the is12Hour check). This ensures that the fix is released as soon as possible, and is not delayed by any further discussion around the date format.

Please review the PR and let's don't delay it.

> we should open a separate issue to discuss the addition of a "date format" prop. If you'd like to bring the topic up during the next meeting, we'll be able to link to that issue for folks to chime in and discuss

Agree with you, I will create a new issue.

File: packages/components/src/date-time/time.js [hunk 1]
```diff
@@ -194,12 +194,7 @@ export function TimePicker( { is12Hour, currentTime, onChange } ) {
 		</div>
 	);
 
-	const dayMonthFormat = is12Hour ? (
```
Comment (ciampo) [inline packages/components/src/date-time/time.js:197]: In the light of the above, I would simply invert the logic of this check:

```js
const dayMonthFormat = is12Hour ? (
	<>
		{ monthFormat }
		{ dayFormat }
	</>
) : (
	<>
		{ dayFormat }
		{ monthFormat }
	</>
);
```

**Review by ciampo — COMMENTED** [review]
> Please review the PR and let's don't delay it.

Sure! As mentioned above, this PR should simply invert the `is12Hour` check (I've added an inline comment to explain that more explicitely).

We should then:

- adjust the unit tests accordingly
- add a [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) entry
- potentially add more details to the `is12Hour` prop description, to add more clarity around the fact that the date format is `DD-MM-YYYY` when that prop is `false`

**Comment by ciampo** [comment]
> Agree with you, I will create a new issue.

Thank you! It'd be great if you linked it back here once you've created it :)

File: packages/components/src/date-time/time.js (hunk 1 above)
Comment (amustaque97) [reply packages/components/src/date-time/time.js:197]: I have made the changes.

**Comment by amustaque97** [comment]
I have addressed review comments.

**Comment by amustaque97** [comment]
I have created an [issue](https://github.com/WordPress/gutenberg/issues/38221) to introduce new props

**Comment by amustaque97** [comment]
Please don't review PR. I need a day time to review it myself thoroughly. I'm feeling that I have done something wrong. It's late here (going to bed)! Will check it tomorrow. 
Thanks

**Comment by amustaque97** [comment]
cc @ciampo and @paaljoachim 

Feel free to review PR

File: packages/components/CHANGELOG.md [hunk 2]
```diff
@@ -46,6 +46,7 @@
 -   Normalized label line-height and spacing within the `ToolsPanel` ([36387](https://github.com/WordPress/gutenberg/pull/36387))
 -   Remove unused `reakit-utils` from peer dependencies ([#37369](https://github.com/WordPress/gutenberg/pull/37369)).
 -   Update all Emotion dependencies to the latest version to ensure they work correctly with React types ([#37365](https://github.com/WordPress/gutenberg/pull/37365)).
+-   Revert `is12Hour` condition to show `MM-DD-YYYY` when set to `true` ([#37465](https://github.com/WordPress/gutenberg/pull/37465))
```
Comment (ciampo) [inline packages/components/CHANGELOG.md]: Let's not use the term "revert", as we're not undoing any changes — maybe something like

```suggestion
-   `DateTimePicker`: Fix the date format associated to the `is12Hour` prop ([#37465](https://github.com/WordPress/gutenberg/pull/37465))
```

File: packages/components/src/date-time/README.md [hunk 3]
```diff
@@ -52,7 +52,7 @@ The function called when a new date or time has been selected. It is passed the
 
 ### is12Hour
 
-Whether we use a 12-hour clock. With a 12-hour clock, an AM/PM widget is displayed and the time format is assumed to be MM-DD-YYYY.
+Whether we use a 12-hour clock. With a 12-hour clock, an AM/PM widget is displayed and the time format is assumed to be MM-DD-YYYY. If its value is `false`, format DD-MM-YYYY will be used.
```
Comment (ciampo) [inline packages/components/src/date-time/README.md]: Nit:

```suggestion
Whether we use a 12-hour clock. With a 12-hour clock, an AM/PM widget is displayed and the time format is assumed to be `MM-DD-YYYY` (as opposed to the default format `DD-MM-YYYY`).
```

**Review by ciampo — APPROVED** [review]
Code changes LGTM, the date format is now aligned with the docs

| default | `is12Hour = true` |
|---|---|
| Screenshot 2022-01-26 at 10 11 51 |  Screenshot 2022-01-26 at 10 11 59 |


Just left a couple of minor inline comments, once they're addressed we can merge this PR! Thank you @amustaque97

**Comment by ciampo** [comment]
Some e2e tests are failing, likely unrelated to the changes in this PR.

@amustaque97 , would you mind rebasing on top of latest `trunk`? It may help with the e2e tests

**Pull request merged**